### DISCUSSION TITLE
Setting clearing script from the gcode, lifting prints off bed

### DISCUSSION THREAD

**Discussion by ferm10n** [smartin015/continuousprint#136] [Show and tell]
I love this plugin so much. One thing that was a bit annoying to deal with in my use case was creating a clearing script that worked for everything I wanted to print. I've modified my printer to have a "hook" on the extruder assembly that I can use to lift parts off of the bed. I also modify my parts to print with an easily detachable tab that can be hooked into.

![image](https://user-images.githubusercontent.com/7970987/196550380-57294506-7213-46f9-94fe-c3b796b5f8bd.png)

Here's my method for how I got the ability to define a clearing script in my prusa slicer's 3mf file, like this!
![image](https://user-images.githubusercontent.com/7970987/196549335-ca094132-f18e-427b-9da6-59d2d1ff8b5e.png)

Basically I have octoprint configured to run a python script when a new print is started. If the gcode has `; CLEARING_SCRIPT: ` lines in it anywhere, it will use those to build the clearing script!! If none are found, I load a default clearing script which just beeps and does `@pause`. The resulting clearing script is sent to the octoprint api.

Some configs

![image](https://user-images.githubusercontent.com/7970987/196548241-dbda8922-1c5e-4d49-9d08-e8e76c9432b4.png)

![image](https://user-images.githubusercontent.com/7970987/196548287-733807de-1347-4255-919d-f4e4fcbe1931.png)

Here's the PrintStarted script:

```python
#!/usr/bin/env python3
import requests
import os
import sys
import re

octoprintOrigin = os.getenv('OCTOPRINT_ORIGIN', default='http://localhost:5000')
apiKey = os.getenv('OCTOPRINT_API_KEY', default='...')
gcodeFilename = '/octoprint/octoprint/uploads/' + sys.argv[1] # I'm using dockerized octoprint, so /octoprint/octoprint/uploads is where things gets uploaded for me

# We don't want to do anything if it's a continuous printing script
if re.search(r'cp_.*_script', gcodeFilename):
  print('Not processing a continuous print script')
  sys.exit()

# Read the gcode file that was just started.
print('trying to read ' + gcodeFilename)
clearingScript = []
with open (gcodeFilename, 'rt') as file:
  for line in file:
    if line.startswith('; CLEARING_SCRIPT: '):
      clearingScript.append(line.replace('; CLEARING_SCRIPT: ', '').replace('\n', ''))
if len(clearingScript) == 0:
  # load default
  defaultClearingScript = open('/octoprint/octoprint/defaultClearingScript.gcode', 'rt')
  clearingScript = defaultClearingScript.read()
  defaultClearingScript.close()
  print('using default clearing script')
else:
  clearingScript = '\n'.join(clearingScript)
  print('clearing script:')
  print(clearingScript)

r = requests.post(octoprintOrigin + '/api/settings', json={
  "plugins": {
    "continuousprint": {
      "cp_bed_clearing_script": clearingScript
    }
  }
}, headers={"X-Api-Key": apiKey}).raise_for_status()
```

**Comment by smartin015** [comment]
That is so rad, there's some really useful tricks in here... I hadn't thought to use octoprint's event manager to loop back around and autogenerate a clearing script.

[Such elegance!](https://www.youtube.com/watch?v=G9mBFBW__qc)
